backends/dpdk/printUtils.cpp

```diff
@@ -0,0 +1,125 @@
+/* Copyright 2021 Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "printUtils.h"
+
+namespace DPDK {
+// this function takes different subclass of Expression and translate it into
+// string in desired format. For example, for PathExpression, it returns
+// PathExpression->path->name For Member, it returns
+// toStr(Member->expr).Member->member
+cstring toStr(const IR::Expression *const);
+
+// this function takes different subclass of Type and translate it into string
+// in desired format. For example, for Type_Boolean, it returns bool For
+// Type_Bits, it returns bit_<bit_width>
+cstring toStr(const IR::Type *const);
+
+// this function takes different subclass of PropertyValue and translate it into
+// string in desired format. For example, for ExpressionValue, it returns
+// toStr(ExpressionValue->expression)
+cstring toStr(const IR::PropertyValue *const);
+
+cstring toStr(const IR::Constant *const c) {
+    std::ostringstream out;
+    out << "0x" << std::hex << c->value;
+    return out.str();
+}
+
+cstring toDecimal(const IR::Constant*const c) {
+    std::ostringstream out;
+    out << c->value;
+    return out.str();
+}
+
+cstring toStr(const IR::BoolLiteral *const b) {
+    std::ostringstream out;
+    out << b->value;
+    return out.str();
+}
+
+cstring toStr(const IR::Member *const m) {
+    std::ostringstream out;
+    out << m->member.originalName;
+    return toStr(m->expr) + "." + out.str();
+}
+
+cstring toStr(const IR::PathExpression *const p) { return p->path->name; }
+
+cstring toStr(const IR::TypeNameExpression *const p) {
+    return p->typeName->path->name;
+}
+
+cstring toStr(const IR::MethodCallExpression *const m) {
+    if (auto path = m->method->to<IR::PathExpression>()) {
+        return path->path->name.toString();
+    } else {
+        ::error("%1% is not a PathExpression", m->toString());
+    }
+    return "";
+}
+
+cstring toStr(const IR::Expression *const exp) {
```

Comment:
Couldn't this be done nicer using a visitor?

Reply:
Sorry, I don't think I follow this comment completely. Could you please elaborate a little?

Reply:
A visitor class that has a preorder method for each kind of expression could achieve the same result in a nicer way.

Reply:
Sorry about my naïve question.
I think I still don't understand this correctly. These are helper routines for printing. If we make it a visitor how would we return a string there and how the call would be replaced with a visitor function?
For Example here:
https://github.com/p4lang/p4c/blob/4b5f7b43b17d6a6db7a994e2f41a4f6177a6231e/backends/dpdk/spec.cpp#L55

Reply:
The visitor has a stringbuilder that it appends strings to. An Inspector is enough. Look at the ebpf backend for an example.

Reply:
Can we take this refactoring as a seperate PR as it would impact .spec file generation as well? 

This is existing code for .spec file generation. I just moved it from _backend/dpdk/spec.cpp_ to this new file _backend/dpdk/printUtils.cpp_ for reusing some of the print routines in context JSON file generation.

Perhaps, I can file an issue and work on this refactoring after this context JSON related PR?

Reply:
This is called "Technical debt". If you indeed plan to make these changes that's fine, but we've had contributions in the past which were not followed by the promised cleanups. 
We can merge this part of the code as it is, it's not terrible.
But if I am right that a visitor would solve this nicely, it should be actually quite a quick fix. You would essentially define a visitor class, and convert each of these cases of the 'if' into a preorder method. That's it.

Reply:
Fixed.